File: conf/paths.go
// Copyright 2022 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

//go:build !local
// +build !local

package conf

import (
	"os"
	"path"
)

const (
	BrokenArtifactSuffix = "_INCONSISTENT"
)

func getenv(key, fallback string) string {
	value := os.Getenv(key)
	if len(value) == 0 {
		return fallback
	}
	return value
}

var (
	// needed so that we can override it when testing or deploying on partially read-only systems
	DefaultPathConfDir = getenv("MENDER_CONF_DIR", "/etc/mender")
	DefaultPathDataDir = getenv("MENDER_DATA_DIR", "/usr/share/mender")
	DefaultDataStore   = getenv("MENDER_DATASTORE_DIR", "/var/lib/mender")
	DefaultKeyFile     = "mender-agent.pem"

	DefaultConfFile         = path.Join(GetConfDirPath(), "mender.conf")
	DefaultFallbackConfFile = path.Join(GetStateDirPath(), "mender.conf")
)

var (
	// device specific paths
	DefaultArtifactInfoFile  = path.Join(GetConfDirPath(), "artifact_info")
	DefaultArtScriptsPath    = path.Join(GetStateDirPath(), "scripts")
	DefaultRootfsScriptsPath = path.Join(GetConfDirPath(), "scripts")
	DefaultModulesPath       = path.Join(GetDataDirPath(), "modules", "v3")
	DefaultModulesWorkPath   = path.Join(GetStateDirPath(), "modules", "v3")
)

func GetDataDirPath() string {
	return DefaultPathDataDir
}

func GetStateDirPath() string {
	return DefaultDataStore
}

func GetConfDirPath() string {
	return DefaultPathConfDir
}

File: common/conf/platform/posix/paths.cpp
// Copyright 2023 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

#include <common/conf/paths.hpp>

#include <string>

#include <common/conf.hpp>
#include <common/path.hpp>

namespace mender {
namespace common {
namespace conf {
namespace paths {

using namespace std;
namespace conf = mender::common::conf;
namespace path = mender::common::path;

const string DefaultPathConfDir = conf::GetEnv("MENDER_CONF_DIR", path::Join("/etc", "mender"));
const string DefaultPathDataDir =
	conf::GetEnv("MENDER_DATA_DIR", path::Join("/usr/share", "mender"));
const string DefaultDataStore =
	conf::GetEnv("MENDER_DATASTORE_DIR", path::Join("/var/lib", "mender"));
const string DefaultKeyFile = "mender-agent.pem";

const string DefaultConfFile = path::Join(DefaultPathConfDir, "mender.conf");
const string DefaultFallbackConfFile = path::Join(DefaultDataStore, "mender.conf");

// device specific paths
const string DefaultArtScriptsPath = path::Join(DefaultDataStore, "scripts");
const string DefaultRootfsScriptsPath = path::Join(DefaultPathConfDir, "scripts");
const string DefaultModulesPath = path::Join(DefaultPathDataDir, "modules/v3");
const string DefaultModulesWorkPath = path::Join(DefaultDataStore, "modules/v3");
const string DefaultBootstrapArtifactFile = path::Join(DefaultDataStore, "bootstrap.mender");

} // namespace paths
} // namespace conf
} // namespace common
} // namespace mender
